### PR TITLE
Rework fix for stale data in synthetic source to improve performance

### PR DESCRIPTION
**Pull request by lkts**
This PR reworks #112173. Original PR is correct but benchmarks show a significant performance regression when synthesizing source in presence of hundreds of stored field. This is cased by a `SourceLoader.SyntheticFieldLoader.StoredFieldLoader#advanceToDoc` call that was executed for every stored field loader every time source was synthesized, even if the stored field was always empty.

This PR changes the approach to only perform reset of synthetic field loader only when it is known that it is needed, specifically when values were loaded but not used due to presence of ignored source values.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/IgnoreMalformedStoredValues.java [hunk 1]
```diff
@@ -79,6 +79,8 @@ public static IgnoreMalformedStoredValues stored(String fieldName) {
      */
     public abstract void write(XContentBuilder b) throws IOException;
 
+    public abstract void reset();
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/IgnoreMalformedStoredValues.java:82]: Let's add a javadoc comment.

**Comment by kkrik-es** [comment]
Nice, seems simpler. There's a fine line between calling the `softReset` vs `reset`, maybe document the use of softReset where applicable, or the possible need for it in the interface?

Have we tested its perf impact?

**Review by kkrik-es — APPROVED** [review]

**Review by martijnvg — APPROVED** [review]
👍 

I checked and the time spent in `org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(LeafStoredFieldLoader, int, XContentBuilder)` method has been reduced from 60% to 26% with this change. I think this should undo the observed regression.

**Comment by lkts** [comment]
Yes, i verified performance impact using local CPU sampling.

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[](https://github.com/elastic/elasticsearch/pull/112524)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**